vowpalwabbit/lda_core.cc

```diff
@@ -627,48 +626,47 @@ size_t next_pow2(size_t x)
 void save_load(lda &l, io_buf &model_file, bool read, bool text)
 { vw *all = l.all;
   uint64_t length = (uint64_t)1 << all->num_bits;
-  uint64_t stride = (uint64_t)1 << all->reg.stride_shift;
-
+  weight_vector& weights = all->wv;
   if (read)
   { initialize_regressor(*all);
-    for (uint64_t j = 0; j < stride * length; j += stride)
-    { for (size_t k = 0; k < all->lda; k++)
-      { if (all->random_weights)
-        { all->reg.weight_vector[j + k] = (float)(-log(frand48() + 1e-6) + 1.0f);
-          all->reg.weight_vector[j + k] *= (float)(l.lda_D / all->lda / all->length() * 200);
-        }
-      }
-      all->reg.weight_vector[j + all->lda] = all->initial_t;
-    }
+	weight_vector::iterator j = weights.begin(0);
+	weight_vector::iterator w_lda = weights.begin(all->lda);
+	for (; j != weights.end(); ++j, ++w_lda)
+	{ for (weight_vector::iterator::w_iter k = j.begin(); k != j.end(all->lda); ++k)
```

Comment:
A couple of things:

1) Why are the begin iterators almost always defined outside of the for() loop preamble?  I can't see why you just don't do:
`for(iterator jj = weights.begin, kk = weights.begin(all->lda), end = weights.end(all->lda); ...) {}`

Also note that I moved the `end(all->lda)` into the initialization section.  The condition `k != ...` needs to be evaluated on each iteration and it's dubious whether or not the compiler will notice that it's not changing.  So unless the contents of your iterator range is changing during the iteration, it's better to evaluate it just once.  This happens in a lot of places.